src/back/msl/writer.rs

```diff
@@ -786,20 +786,41 @@ impl<W: Write> Writer<W> {
 
     fn put_atomic_fetch(
         &mut self,
-        pointer: Handle<crate::Expression>,
+        expr_handle: Handle<crate::Expression>,
         key: &str,
         value: Handle<crate::Expression>,
         context: &ExpressionContext,
     ) -> BackendResult {
-        write!(
-            self.out,
-            "{}::atomic_fetch_{}_explicit({}",
-            NAMESPACE, key, ATOMIC_REFERENCE
-        )?;
-        self.put_expression(pointer, context, true)?;
-        write!(self.out, ", ")?;
-        self.put_expression(value, context, true)?;
-        write!(self.out, ", {}::memory_order_relaxed)", NAMESPACE)?;
+        // Atomic ops are structured differently from a normal load.
+        // Rewrite this for the bounds-check case so that the bounds check happens once
+        // when resolving the ptr, and then use an unchecked access within the atomic expression itself.
+        let policy = context.choose_bounds_check_policy(expr_handle);
+
+        let put_unchecked_atomic_access = |writer: &mut Writer<W>| -> BackendResult {
+            write!(
+                writer.out,
+                "{}::atomic_fetch_{}_explicit({}",
+                NAMESPACE, key, ATOMIC_REFERENCE
+            )?;
+            writer.put_access_chain(expr_handle, policy, context)?;
+            write!(writer.out, ", ")?;
+            writer.put_expression(value, context, true)?;
+            write!(writer.out, ", {}::memory_order_relaxed)", NAMESPACE)?;
+            Ok(())
+        };
```

Comment:
Yes, I noticed the duplicated code as well. Naga's style is to not use closures in this way, to keep control flow simple. This is certainly subjective, but it's what we've settled on.

Instead, please declare a `bool`, and then have two `if` statements on either side of the common code.